perception/autoware_tensorrt_bevdet/src/gatherbev_plugin.cu

```diff
@@ -0,0 +1,308 @@
+{
+  return inputTypes[0];  // 与adj_feat一致
+}
+
+DimsExprs GatherBEVPlugin::getOutputDimensions(
+  int32_t outputIndex, const DimsExprs * inputs, int32_t nbInputs,
+  IExprBuilder & exprBuilder) noexcept
+{
+  // input[0] == adj_feat   b x 8 x 80 x 128 x 128
+  // input[1] == curr_feat  b x 80 x 128 x 128
+  // input[2] == flag       b x 1
+  // out[0]                 b x (8+1)*80 x 128 x 128
+  DimsExprs ret;
+  ret.nbDims = inputs[0].nbDims - 1;
+
+  IDimensionExpr const * n_feat =
+    exprBuilder.operation(DimensionOperation::kSUM, *inputs[0].d[1], *exprBuilder.constant(1));
+  ret.d[0] = inputs[0].d[0];
+  ret.d[1] = exprBuilder.operation(DimensionOperation::kPROD, *n_feat, *inputs[0].d[2]);
+  ret.d[2] = inputs[0].d[3];
+  ret.d[3] = inputs[0].d[4];
+
+  return ret;
+}
+
+bool GatherBEVPlugin::supportsFormatCombination(
+  int32_t pos, const PluginTensorDesc * inOut, int32_t nbInputs, int32_t nbOutputs) noexcept
+{
+  // adj_feat
+  if (pos == 0) {
+    return (inOut[pos].type == DataType::kFLOAT || inOut[pos].type == DataType::kHALF) &&
+           inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 1) {  // curr_feat
+    return inOut[0].type == inOut[1].type && inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 3) {  // out
+    return inOut[0].type == inOut[3].type && inOut[pos].format == TensorFormat::kLINEAR;
+  } else if (pos == 2) {
+    return inOut[pos].type == DataType::kINT32 && inOut[pos].format == TensorFormat::kLINEAR;
+  }
+  return false;
+}
+
+void GatherBEVPlugin::configurePlugin(
+  const DynamicPluginTensorDesc * in, int32_t nbInputs, const DynamicPluginTensorDesc * out,
+  int32_t nbOutputs) noexcept
+{
+  return;
+}
+
+size_t GatherBEVPlugin::getWorkspaceSize(
+  const PluginTensorDesc * inputs, int32_t nbInputs, const PluginTensorDesc * outputs,
+  int32_t nbOutputs) const noexcept
+{
+  return 0;
+}
+
+int32_t GatherBEVPlugin::enqueue(
+  const PluginTensorDesc * inputDesc, const PluginTensorDesc * outputDesc,
+  const void * const * inputs, void * const * outputs, void * workspace,
+  cudaStream_t stream) noexcept
+{
+  // input[0] == adj_feat   b x 8 x 80 x 128 x 128
+  // input[1] == curr_feat  b x 80 x 128 x 128
+  // input[2] == flag       b x 1
+  // out[0]                 b x (8+1)*80 x 128 x 128
+
+  // int nthreads, // b * (adj_num + 1) * map_size
+  // int adj_num,
+  // int channel,
+  // int map_size,
+
+  // int flag = 0;
+  // CHECK_CUDA(cudaMemcpy(&flag, inputs[2], sizeof(int), cudaMemcpyDeviceToHost));
+
+  int b = inputDesc[0].dims.d[0];
+  int adj_num = inputDesc[0].dims.d[1];
+  int map_size = inputDesc[0].dims.d[3] * inputDesc[0].dims.d[4];
+  int channel = inputDesc[0].dims.d[2];
+
+  int feat_step = inputDesc[1].dims.d[1] * inputDesc[1].dims.d[2] * inputDesc[1].dims.d[3];
+
+  int nthreads = b * (adj_num + 1) * map_size;
+
+  dim3 grid(GET_BLOCKS(nthreads));
+  dim3 block(NUM_THREADS);
+
+  switch (int(outputDesc[0].type)) {
```

Comment:
The switch statement for handling different data types could be simplified and potentially optimized. Consider using template metaprogramming or function overloading instead of a switch statement to avoid branching and increase code clarity. This would allow the compiler to perform better optimizations.